replace cluster-id with new supervisor-id with CSI full sync

divyenpatel opened the pull request:
**What this PR does / why we need it**:

- delete volume metadata with the old `cluster-id` and add the same volume metadata back with the new `supervisor-id`.
- removed previous logic to replace  `config.Global.ClusterID` with  `config.Global.SupervisorID` and using `config.Global.ClusterID` to push metadata using new SupervisorID
- updated other supervisor services to use the new `SupervisorID` to push volume metadata.


**Testing done**:
[Test logs](https://gist.github.com/divyenpatel/e408cf2faaa31914bf9a899f974aa2db)

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
replace cluster-id with new supervisor-id with CSI full sync
```

lintongj commented on pkg/syncer/cnsoperator/controller/cnsvolumemetadata/cnsvolumemetadata_controller.go:
```diff
@@ -329,7 +330,11 @@ func (r *ReconcileCnsVolumeMetadata) updateCnsMetadata(ctx context.Context,
 		if instance.Spec.EntityType == cnsv1alpha1.CnsOperatorEntityTypePV {
-			clusterid = r.configInfo.Cfg.Global.ClusterID
+			if common.UseSuperVisorIDasClusterID {
+				clusterid = r.configInfo.Cfg.Global.SupervisorID
```
nit. clusterid -> clusterID or clusterId?

divyenpatel replied:
done

lintongj reviewed (commented):
Generally looks good except some nits.

divyenpatel commented on pkg/csi/service/common/vsphereutil.go:
```diff
@@ -32,7 +32,6 @@ import (
 	"golang.org/x/net/context"
 	"google.golang.org/grpc/codes"
 	"google.golang.org/protobuf/types/known/timestamppb"
-
```
added back new line.

divyenpatel commented on pkg/csi/service/common/vsphereutil.go:
```diff
@@ -173,7 +172,11 @@ func CreateBlockVolumeUtil(ctx context.Context, clusterFlavor cnstypes.CnsCluste
 	}
 	var containerClusterArray []cnstypes.CnsContainerCluster
-	containerCluster := vsphere.GetContainerCluster(manager.CnsConfig.Global.ClusterID,
+	cluserID := manager.CnsConfig.Global.ClusterID
```
done.

divyenpatel commented on pkg/csi/service/common/vsphereutil.go:
```diff
@@ -327,8 +330,12 @@ func CreateFileVolumeUtil(ctx context.Context, clusterFlavor cnstypes.CnsCluster
 		})
 	}
 
+	cluserID := manager.CnsConfig.Global.ClusterID
```
done

divyenpatel commented on pkg/syncer/cnsoperator/controller/cnsregistervolume/util.go:
```diff
@@ -73,7 +73,13 @@ func constructCreateSpecForInstance(r *ReconcileCnsRegisterVolume,
-	containerCluster := vsphere.GetContainerCluster(r.configInfo.Cfg.Global.ClusterID,
+	var clusterIForVolumeMetadata string
+	if common.UseSuperVisorIDasClusterID {
+		clusterIForVolumeMetadata = r.configInfo.Cfg.Global.SupervisorID
```
done

divyenpatel commented on pkg/csi/service/wcp/controller.go:
```diff
@@ -89,13 +89,10 @@ func (c *controller) Init(config *cnsconfig.Config, version string) error {
+			if config.Global.SupervisorID == "" {
 				return logger.LogNewError(log, "supervisor-id is not set in the vsphere-config-secret")
 			}
+			common.UseSuperVisorIDasClusterID = true
```
removed usage of global variable `common.UseSuperVisorIDasClusterID`, and supplied pre-fetched status of tkgs-ha feature gate to determine if we need to use supervisorID or clusterID from the config secret.

divyenpatel commented:
@shalini-b @lintongj I have addressed your review comments here - https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/1682/commits/968ec3ef22e9427fe3dc2b17f3a32fd116070bb1

shalini-b commented on pkg/csi/service/wcp/controller.go:
```diff
@@ -511,8 +496,9 @@ func (c *controller) createBlockVolume(ctx context.Context, req *csi.CreateVolum
 		VsanDirectDatastoreURL: selectedDatastoreURL,
 	}
 	candidateDatastores := append(sharedDatastores, vsanDirectDatastores...)
+	useSupervisorId := commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx, common.TKGsHA)
```
Do you want to set `useSupervisorId` in `Init` and use that in every CreateVolume call?

divyenpatel replied:
utilized the check we used for making changes in the CreateVolume Call

```
	isTKGSHAEnabled := commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx, common.TKGsHA)
	if isTKGSHAEnabled {
		// TKGS-HA feature is enabled
		// Identify the topology keys in Accessibility requirements.
```

shalini-b commented on pkg/syncer/metadatasyncer.go:
```diff
@@ -182,17 +183,13 @@ func InitMetadataSyncer(ctx context.Context, clusterFlavor cnstypes.CnsClusterFl
 				return err
 			}
 			if len(clusterComputeResourceMoIds) > 0 {
```
Do we still need this check here? If it is workload flavor and TKGs-HA is enabled, we can assign `supervisor-id` to  `clusterIDforVolumeMetadata`

divyenpatel replied:
moved assignment for `clusterIDforVolumeMetadata` out

```
			if len(clusterComputeResourceMoIds) > 0 {
				if configInfo.Cfg.Global.SupervisorID == "" {
					return logger.LogNewError(log, "supervisor-id is not set in the vsphere-config-secret")
				}
			}
			clusterIDforVolumeMetadata = configInfo.Cfg.Global.SupervisorID
```

shalini-b commented on pkg/syncer/fullsync.go:
```diff
@@ -116,6 +116,65 @@ func CsiFullSync(ctx context.Context, metadataSyncer *metadataSyncInformer) erro
 		return err
 	}
 
+	if metadataSyncer.clusterFlavor == cnstypes.CnsClusterFlavorWorkload &&
```
We cannot merge this PR till CNS gets back on how to remove the clusterID from ContainerClusterArray right?

divyenpatel replied:
We can merge CSI side changes as it is guarded with the `tkgs-ha` feature gate.
we need to use the new `supervisor-id` so this change is required.

lintongj commented on pkg/csi/service/wcp/controller.go:
```diff
@@ -641,8 +627,9 @@ func (c *controller) createFileVolume(ctx context.Context, req *csi.CreateVolume
 			"no datastores found to create file volume")
 	}
 	filterSuspendedDatastores := commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx, common.CnsMgrSuspendCreateVolume)
+	useSupervisorId := commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx, common.TKGsHA)
```
would be better to name it as `isTKGSHAEnabled`?

Also, would it be better to pass a func rather than a value here? This way, when the `isTKGSHAEnabled` func is called, the FSS state in the real-time will be used. Instead, passing a value might end up with a stale  one.

divyenpatel replied:
used `isTKGSHAEnabled` instead of `useSupervisorId`.
it is ok to rely on the stale value of `isTKGSHAEnabled` as we do not support runtime feature state enablement. checking feature state requires checking config-map and trip to the API server so using cached value.

lintongj reviewed (approved):
/approve

shalini-b commented on pkg/syncer/fullsync.go:
```diff
@@ -116,6 +116,101 @@ func CsiFullSync(ctx context.Context, metadataSyncer *metadataSyncInformer) erro
+					if entityMetadata.GetCnsEntityMetadata().ClusterID ==
+						metadataSyncer.configInfo.Cfg.Global.ClusterID {
+						// Delete metadata for associated with old cluster ID
+						oldk8sEntityMetadata := entityMetadata.(*cnstypes.CnsKubernetesEntityMetadata)
```
Didn't we decide to dereference the `entityMetadata` object here?

shalini-b commented on pkg/syncer/fullsync.go:
```diff
@@ -116,6 +116,101 @@ func CsiFullSync(ctx context.Context, metadataSyncer *metadataSyncInformer) erro
+						Id: volume.VolumeId.Id,
+					},
+					Metadata: cnstypes.CnsVolumeMetadata{
+						ContainerCluster:      volume.Metadata.ContainerCluster,
```
Can't we set the value to `metadataSyncer.configInfo.Cfg.Global.SupervisorID` directly?

divyenpatel replied:
done

shalini-b commented on pkg/syncer/fullsync.go:
```diff
@@ -116,6 +116,101 @@ func CsiFullSync(ctx context.Context, metadataSyncer *metadataSyncInformer) erro
+				log.Debugf("Calling UpdateVolumeMetadata for volume %s with updateSpec: %+v",
+					updateSpec.VolumeId.Id, spew.Sdump(updateSpec))
+				if err := metadataSyncer.volumeManager.UpdateVolumeMetadata(ctx, &updateSpec); err != nil {
+					log.Warnf("FullSync Replace ClusterID with new SupervisorID :UpdateVolumeMetadata failed with err %v", err)
```
nit: `log.Warnf("FullSync Replace ClusterID with new SupervisorID :UpdateVolumeMetadata failed with err %v", err)` -> `log.Warnf("FullSync: UpdateVolumeMetadata failed while replacing clusterID with supervisorID. Error: %+v", err)`

divyenpatel replied:
done

shalini-b commented:
/ok-to-test

BaluDontu commented:
Hey @divyenpatel Why did't we run Vanilla tests block/file pipeline runs for this?
